File: echo-bot/ruby/echo_bot.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

##
# SendSeven API - Echo Bot Example (Ruby/Sinatra)
#
# A simple bot that automatically replies to incoming messages.

require 'sinatra'
require 'json'
require 'net/http'
require 'uri'
require 'openssl'

# Load .env file
def load_env_file
  env_file = File.join(__dir__, '.env')
  return unless File.exist?(env_file)

  File.readlines(env_file).each do |line|
    line = line.strip
    next if line.empty? || line.start_with?('#')

    key, value = line.split('=', 2)
    ENV[key.strip] ||= value.strip if key && value
  end
end

load_env_file

# Configuration
API_TOKEN = ENV['SENDSEVEN_API_TOKEN'] || ''
TENANT_ID = ENV['SENDSEVEN_TENANT_ID'] || ''
API_URL = ENV['SENDSEVEN_API_URL'] || 'https://api.sendseven.com/api/v1'
WEBHOOK_SECRET = ENV['WEBHOOK_SECRET'] || ''
PORT = (ENV['PORT'] || '3000').to_i

set :port, PORT
set :bind, '0.0.0.0'

# Validate configuration
if API_TOKEN.empty?
  puts 'Error: SENDSEVEN_API_TOKEN environment variable is required'
  exit 1
end

if TENANT_ID.empty?
  puts 'Error: SENDSEVEN_TENANT_ID environment variable is required'
  exit 1
end

if WEBHOOK_SECRET.empty?
  puts 'Warning: WEBHOOK_SECRET not set - signatures will not be verified!'
end

# Track processed deliveries (use Redis/database in production)
$processed_deliveries = {}

##
# Verify the webhook signature using HMAC-SHA256.
def verify_signature(payload, signature, timestamp)
  return false unless signature.start_with?('sha256=')

  provided_sig = signature[7..] # Remove 'sha256=' prefix

  # Reconstruct message
  message = "#{timestamp}.#{payload}"

  # Compute expected signature
  expected_sig = OpenSSL::HMAC.hexdigest('SHA256', WEBHOOK_SECRET, message)

  # Timing-safe comparison
  secure_compare(expected_sig, provided_sig)
end

##
# Timing-safe string comparison
def secure_compare(a, b)
  return false unless a.bytesize == b.bytesize

  l = a.unpack('C*')
  res = 0
  b.each_byte { |byte| res |= byte ^ l.shift }
  res.zero?
end

##
# Send a reply message to a conversation.
def send_reply(conversation_id, text)
  uri = URI("#{API_URL}/messages")

  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = uri.scheme == 'https'

  request = Net::HTTP::Post.new(uri)
  request['Authorization'] = "Bearer #{API_TOKEN}"
  request['X-Tenant-ID'] = TENANT_ID
  request['Content-Type'] = 'application/json'

  request.body = {
    conversation_id: conversation_id,
    text: text,
    message_type: 'text'
  }.to_json

  response = http.request(request)

  raise "API Error #{response.code}: #{response.body}" if response.code.to_i >= 400

  JSON.parse(response.body)
end

##
# Generate a reply based on message type.
def generate_reply(message_type, message_text)
  case message_type
  when 'text'
    message_text.empty? ? 'I received your message!' : "You said: \"#{message_text}\""
  when 'image'
    'I received your image! 📷'
  when 'audio'
    'I received your audio message! 🎵'
  when 'video'
    'I received your video! 🎬'
  when 'document'
    'I received your document! 📄'
  else
    'I received your message!'
  end
end

##
# Handle incoming webhooks
post '/webhooks/sendseven' do
  content_type :json

  # Get headers
  signature = request.env['HTTP_X_SENDSEVEN_SIGNATURE'] || ''
  timestamp = request.env['HTTP_X_SENDSEVEN_TIMESTAMP'] || ''
  delivery_id = request.env['HTTP_X_SENDSEVEN_DELIVERY_ID'] || ''

  # Verify required headers
  if signature.empty? || timestamp.empty? || delivery_id.empty?
    puts 'Missing required webhook headers'
    halt 400, { error: 'Missing required headers' }.to_json
  end

  # Check for duplicate (idempotency)
  if $processed_deliveries[delivery_id]
    puts "Duplicate delivery #{delivery_id}, skipping"
    return { success: true, duplicate: true }.to_json
  end

  # Read body
  payload = request.body.read

  # Verify signature
  if !WEBHOOK_SECRET.empty? && !verify_signature(payload, signature, timestamp)
    puts "Invalid signature for delivery #{delivery_id}"
    halt 401, { error: 'Invalid signature' }.to_json
  end

  # Parse payload
  begin
    data = JSON.parse(payload)
  rescue JSON::ParserError
    halt 400, { error: 'Invalid JSON' }.to_json
  end

  event_type = data['type'] || ''

  # Only process message.received events
  unless event_type == 'message.received'
    return { success: true, skipped: true }.to_json
  end

  # Extract message details
  message = data.dig('data', 'message') || {}
  contact = data.dig('data', 'contact') || {}

  # Only respond to inbound messages (avoid loops)
  unless message['direction'] == 'inbound'
    return { success: true, skipped: 'outbound' }.to_json
  end

  conversation_id = message['conversation_id'] || ''
  message_type = message['message_type'] || 'text'
  message_text = message['text'] || ''
  contact_name = contact['name'] || 'there'

  preview = message_text[0, 50] || '[media]'
  preview = '[media]' if preview.empty?
  puts "Received message from #{contact_name}: #{preview}"

  # Generate and send reply
  reply_text = generate_reply(message_type, message_text)

  begin
    result = send_reply(conversation_id, reply_text)
    puts "Reply sent: #{result['id']}"
    $processed_deliveries[delivery_id] = true
  rescue StandardError => e
    puts "Failed to send reply: #{e.message}"
  end

  # Return 200 OK
  { success: true }.to_json
end

##
# Health check endpoint
get '/health' do
  content_type :json
  { status: 'ok' }.to_json
end

puts "Echo Bot listening on port #{PORT}"
puts "Webhook endpoint: http://localhost:#{PORT}/webhooks/sendseven"

File: echo-bot/python/echo_bot.py
#!/usr/bin/env python3
"""
SendSeven API - Echo Bot Example (Python/Flask)

A simple bot that automatically replies to incoming messages.
"""

import os
import hmac
import hashlib
import json
import requests
from flask import Flask, request, jsonify
from dotenv import load_dotenv

load_dotenv()

app = Flask(__name__)

# Configuration
API_TOKEN = os.getenv("SENDSEVEN_API_TOKEN", "")
TENANT_ID = os.getenv("SENDSEVEN_TENANT_ID", "")
API_URL = os.getenv("SENDSEVEN_API_URL", "https://api.sendseven.com/api/v1")
WEBHOOK_SECRET = os.getenv("WEBHOOK_SECRET", "")
PORT = int(os.getenv("PORT", 3000))

# Track processed delivery IDs to avoid duplicates
processed_deliveries = set()


def verify_signature(payload: bytes, signature: str, timestamp: str) -> bool:
    """Verify the webhook signature using HMAC-SHA256."""
    if not signature.startswith("sha256="):
        return False

    provided_sig = signature[7:]
    payload_dict = json.loads(payload)
    json_payload = json.dumps(payload_dict, separators=(",", ":"), sort_keys=True)
    message = f"{timestamp}.{json_payload}"

    expected_sig = hmac.new(
        WEBHOOK_SECRET.encode("utf-8"),
        message.encode("utf-8"),
        hashlib.sha256
    ).hexdigest()

    return hmac.compare_digest(expected_sig, provided_sig)


def send_reply(conversation_id: str, text: str) -> dict:
    """Send a reply message to a conversation."""
    url = f"{API_URL}/messages"

    headers = {
        "Authorization": f"Bearer {API_TOKEN}",
        "X-Tenant-ID": TENANT_ID,
        "Content-Type": "application/json",
    }

    payload = {
        "conversation_id": conversation_id,
        "text": text,
        "message_type": "text",
    }

    response = requests.post(url, json=payload, headers=headers)
    response.raise_for_status()
    return response.json()


@app.route("/webhooks/sendseven", methods=["POST"])
def handle_webhook():
    """Handle incoming SendSeven webhooks."""
    # Get headers
    signature = request.headers.get("X-Sendseven-Signature", "")
    timestamp = request.headers.get("X-Sendseven-Timestamp", "")
    delivery_id = request.headers.get("X-Sendseven-Delivery-Id", "")

    # Verify required headers
    if not all([signature, timestamp, delivery_id]):
        return jsonify({"error": "Missing required headers"}), 400

    # Check for duplicate delivery (idempotency)
    if delivery_id in processed_deliveries:
        print(f"Duplicate delivery {delivery_id}, skipping")
        return jsonify({"success": True, "duplicate": True}), 200

    # Verify signature
    if WEBHOOK_SECRET and not verify_signature(request.data, signature, timestamp):
        print(f"Invalid signature for delivery {delivery_id}")
        return jsonify({"error": "Invalid signature"}), 401

    # Parse payload
    payload = request.get_json()
    event_type = payload.get("type", "")

    # Only process message.received events
    if event_type != "message.received":
        return jsonify({"success": True, "skipped": True}), 200

    # Extract message details
    data = payload.get("data", {})
    message = data.get("message", {})
    contact = data.get("contact", {})

    # Only respond to inbound messages (avoid loops)
    if message.get("direction") != "inbound":
        return jsonify({"success": True, "skipped": "outbound"}), 200

    conversation_id = message.get("conversation_id")
    message_type = message.get("message_type", "text")
    message_text = message.get("text", "")
    contact_name = contact.get("name", "there")

    print(f"Received message from {contact_name}: {message_text[:50] if message_text else '[media]'}")

    # Generate reply based on message type
    if message_type == "text" and message_text:
        reply_text = f'You said: "{message_text}"'
    elif message_type == "image":
        reply_text = "I received your image! 📷"
    elif message_type == "audio":
        reply_text = "I received your audio message! 🎵"
    elif message_type == "video":
        reply_text = "I received your video! 🎬"
    elif message_type == "document":
        reply_text = "I received your document! 📄"
    else:
        reply_text = "I received your message!"

    # Send the reply
    try:
        result = send_reply(conversation_id, reply_text)
        print(f"Reply sent: {result.get('id')}")
        processed_deliveries.add(delivery_id)
    except requests.HTTPError as e:
        print(f"Failed to send reply: {e.response.status_code} - {e.response.text}")
    except Exception as e:
        print(f"Error sending reply: {e}")

    return jsonify({"success": True}), 200


if __name__ == "__main__":
    # Validate configuration
    if not API_TOKEN:
        print("Error: SENDSEVEN_API_TOKEN is required")
        exit(1)
    if not TENANT_ID:
        print("Error: SENDSEVEN_TENANT_ID is required")
        exit(1)

    if not WEBHOOK_SECRET:
        print("Warning: WEBHOOK_SECRET not set - signatures will not be verified!")

    print(f"Echo Bot starting on port {PORT}")
    print(f"Webhook endpoint: http://localhost:{PORT}/webhooks/sendseven")
    app.run(host="0.0.0.0", port=PORT, debug=True)
